### PR TITLE
MG-2145 - Generate mocks with mockery for Lora service 

### LINKED ISSUE #2145: Feature: Generate mocks with mockery for Lora service
### Is your feature request related to a problem? Please describe.

Requested feature is not related to a problem.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Generate mocks with mockery and rewrite test for Lora service.

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?
This is a feature because its add the following functionality: generates mocks for Lora.

## What does this do?
It generates the mocks for Lora service.

## Which issue(s) does this PR fix/relate to?

- Related Issue #2145 
- Resolves #2145 

## Have you included tests for your changes?
Yes, I have included tests

## Did you document any new/modified feature?
No, I have not updated the documentation.

### CODE REVIEW

File: .github/workflows/check-generated-files.yml [hunk 1]
```diff
@@ -159,3 +161,4 @@ jobs:
           check_mock_changes ./pkg/groups/mocks/service.go "Groups Service ./pkg/groups/mocks/service.go"
           check_mock_changes ./bootstrap/mocks/service.go "Bootstrap Service ./bootstrap/mocks/service.go"
           check_mock_changes ./bootstrap/mocks/configs.go "Bootstrap Repository ./bootstrap/mocks/configs.go"
+          check_mock_changes ./lora/mocks/routes.go "LoRa Repository ./lora/mocks/routes.go"
```
Comment (rodneyosodo) [inline .github/workflows/check-generated-files.yml:181]:
```suggestion
          check_mock_changes ./lora/mocks/routes.go "LoRa Routemap Repository ./lora/mocks/routes.go"
```

File: lora/adapter_test.go [hunk 2]
```diff
@@ -28,17 +30,23 @@ const (
...
 
-	return lora.New(pub, thingsRM, channelsRM, connsRM)
+	return lora.New(pub, thingsRM, channelsRM, connsRM), pub, thingsRM, channelsRM, connsRM
 }
 
 func TestPublish(t *testing.T) {
-	svc := newService()
+	svc, pub, thingsRM, channelsRM, connsRM := newService()
+
+	repoCall := channelsRM.On("Save", context.Background(), mock.Anything, mock.Anything).Return(nil)
```
Comment (rodneyosodo) [inline lora/adapter_test.go]: Try and reduce the usage of `mock.Anything` if you know they parameter being passed

File: lora/adapter_test.go [hunk 3]
```diff
@@ -103,11 +128,20 @@ func TestPublish(t *testing.T) {
...
 			},
+			connectionsErr: lora.ErrNotConnected,
 		},
 	}
 
 	for _, tc := range cases {
+		repoCall := thingsRM.On("Get", context.Background(), mock.Anything).Return("", tc.getThingErr)
+		repoCall1 := channelsRM.On("Get", context.Background(), mock.Anything).Return("", tc.getChannelErr)
+		repoCall2 := connsRM.On("Get", context.Background(), mock.Anything).Return("", tc.connectionsErr)
+		repoCall3 := pub.On("Publish", context.Background(), mock.Anything, mock.Anything).Return(nil)
```
Comment (rodneyosodo) [inline lora/adapter_test.go]: Add `publisherr` for unhappy flows

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: lora/adapter_test.go (hunk 2 above)
Comment (nyagamunene) [reply lora/adapter_test.go]: Okay

**Review by felixgateru — COMMENTED** [review]
Consider opening a ticket to increase test coverage which would involve generating service mocks.

File: lora/adapter_test.go [hunk 4]
```diff
@@ -48,19 +57,32 @@ func TestPublish(t *testing.T) {
...
-		desc string
-		err  error
-		msg  lora.Message
+		desc           string
+		err            error
+		msg            lora.Message
+		getThingErr    error
+		getChannelErr  error
+		connectionsErr error
+		publishErr     error
```
Comment (dborovcanin) [inline lora/adapter_test.go:59]: These errors are mostly not used (i.e. only `publishErr` is checked). All the errors should be asserted and we need cases for different repo call errors.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: .github/workflows/check-generated-files.yml [hunk 5]
```diff
@@ -61,6 +61,7 @@ jobs:
               - "bootstrap/service.go"
               - "bootstrap/configs.go"
               - "invitations/invitations.go"
+              - "lora/routesmap.go"
```
Comment (rodneyosodo) [inline .github/workflows/check-generated-files.yml]:
```suggestion
              - "lora/routemap.go"
```

File: lora/adapter_test.go [hunk 6]
```diff
@@ -28,17 +31,23 @@ const (
...
+	pub := new(pubmocks.PubSub)
+	thingsRM := new(mocks.RouteMapRepository)
+	channelsRM := new(mocks.RouteMapRepository)
+	connsRM := new(mocks.RouteMapRepository)
 
-	return lora.New(pub, thingsRM, channelsRM, connsRM)
+	return lora.New(pub, thingsRM, channelsRM, connsRM), pub, thingsRM, channelsRM, connsRM
 }
 
 func TestPublish(t *testing.T) {
```
Comment (rodneyosodo) [inline lora/adapter_test.go:47]: You can take this opportunity to test other functions, we are only testing `Publish`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: lora/adapter_test.go [hunk 7]
```diff
@@ -103,11 +135,157 @@ func TestPublish(t *testing.T) {
...
+	svc, _, thingsRM, channelsRM, connsRM := newService()
+
+	repoCall := channelsRM.On("Save", context.Background(), chanID, appID).Return(nil)
+	repoCall1 := thingsRM.On("Save", context.Background(), thingID, devEUI).Return(nil)
+
+	err := svc.CreateChannel(context.Background(), chanID, appID)
+	require.Nil(t, err, fmt.Sprintf("unexpected error: %s\n", err))
+
+	err = svc.CreateThing(context.Background(), thingID, devEUI)
+	require.Nil(t, err, fmt.Sprintf("unexpected error: %s\n", err))
```
Comment (felixgateru) [inline lora/adapter_test.go]: Is this section necessary?

File: lora/adapter_test.go [hunk 8]
```diff
@@ -103,11 +135,157 @@ func TestPublish(t *testing.T) {
...
+			channelID:     chanID,
+			thingID:       thingID,
+			getThingErr:   nil,
+			getChannelErr: nil,
+		},
+		{
+			desc:        "connect thing with non existing thing",
+			err:         lora.ErrNotFoundDev,
+			channelID:   chanID,
+			thingID:     "wrong",
```
Comment (felixgateru) [inline lora/adapter_test.go]: Create a global variable such as `invalid` to match all other tests

File: lora/adapter_test.go [hunk 9]
```diff
@@ -103,11 +135,157 @@ func TestPublish(t *testing.T) {
...
+			getThingErr: lora.ErrNotFoundDev,
+		},
+		{
+			desc:          "connect thing with non existing channel",
+			err:           lora.ErrNotFoundApp,
+			channelID:     "wrong",
```
Comment (felixgateru) [inline lora/adapter_test.go]: Same as above

**Review by felixgateru — CHANGES_REQUESTED** [review]

File: lora/adapter_test.go (hunk 7 above)
Comment (nyagamunene) [reply lora/adapter_test.go]: Removed it

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Run and then open `cover.html` on your browser to see the missing cases

```bash
cd lora && go test -tags test -v --race -covermode=atomic -coverprofile=cover.out ./... && go tool cover -html cover.out -o cover.html
```

**Review by rodneyosodo — DISMISSED** [review]
LGTM

File: lora/adapter_test.go [hunk 10]
```diff
@@ -103,11 +109,383 @@ func TestPublish(t *testing.T) {
...
+				DevEUI:        devEUI2,
+				Object:        map[string]interface{}{"key": "value"},
+			},
 		},
 	}
 
 	for _, tc := range cases {
+		repoCall := thingsRM.On("Get", context.Background(), tc.msg.DevEUI).Return(tc.msg.DevEUI, tc.getThingErr)
+		repoCall1 := channelsRM.On("Get", context.Background(), tc.msg.ApplicationID).Return(tc.msg.ApplicationID, tc.getChannelErr)
+		repoCall2 := connsRM.On("Get", context.Background(), mock.Anything).Return(mock.Anything, tc.connectionsErr)
```
Comment (rodneyosodo) [inline lora/adapter_test.go]: Avoid the use of `mock.Anything` especially on return types

File: lora/adapter_test.go [hunk 11]
```diff
@@ -103,11 +109,383 @@ func TestPublish(t *testing.T) {
...
+
+	for _, tc := range cases {
+		repoCall := thingsRM.On("Get", context.Background(), tc.thingID).Return(devEUI, tc.getThingErr)
+		repoCall1 := channelsRM.On("Get", context.Background(), tc.channelID).Return(appID, tc.getChannelErr)
+		repoCall2 := connsRM.On("Remove", context.Background(), mock.Anything).Return(tc.err)
+
+		err := svc.DisconnectThing(context.Background(), tc.channelID, tc.thingID)
+		switch err {
+		case nil:
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
```
Comment (rodneyosodo) [inline lora/adapter_test.go]: No need for this case since if the error is nil it will contain nil

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by rodneyosodo — DISMISSED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]

File: .github/workflows/check-generated-files.yml (hunk 5 above)
Comment (dborovcanin) [reply .github/workflows/check-generated-files.yml]: This one is not fixed.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

File: lora/adapter_test.go [hunk 12]
```diff
@@ -103,11 +109,372 @@ func TestPublish(t *testing.T) {
...
 		},
 	}
 
 	for _, tc := range cases {
+		repoCall := thingsRM.On("Get", context.Background(), tc.msg.DevEUI).Return(tc.msg.DevEUI, tc.getThingErr)
+		repoCall1 := channelsRM.On("Get", context.Background(), tc.msg.ApplicationID).Return(tc.msg.ApplicationID, tc.getChannelErr)
+		repoCall2 := connsRM.On("Get", context.Background(), mock.Anything).Return("", tc.connectionsErr)
+		repoCall3 := pub.On("Publish", context.Background(), tc.msg.ApplicationID, mock.Anything).Return(tc.publishErr)
 		err := svc.Publish(context.Background(), &tc.msg)
+		fmt.Println(err)
```
Comment (WashingtonKK) [inline lora/adapter_test.go:140]: Please remove these debug logs in a different PR since this was merged.

**Pull request merged**